[AArch64] Fix Windows prologue handling to pair more registers.

efriedma-quic opened the pull request:
Currently, there's code to suppress pairing, but we don't actually need to suppress that; we just need to suppress the formation of pre-decrement/post-increment instructions.

Pairing saves an instruction in some cases, and enables packed unwind in some cases.  (There's a comment in the code noting we could enable packed unwind in more cases, but that's not clearly profitable.)

efriedma-quic committed:
[AArch64] Fix Windows prologue handling to pair more registers.

Currently, there's code to suppress pairing, but we don't actually need
to suppress that; we just need to suppress the formation of
pre-decrement/post-increment instructions.

Pairing saves an instruction in some cases, and enables packed unwind in
some cases.  (There's a comment in the code noting we could enable
packed unwind in more cases, but that's not clearly profitable.)

efriedma-quic committed:
Fix test.

compnerd reviewed (approved):
Thanks!

mstorsjo commented on llvm/test/CodeGen/AArch64/wineh-pac.ll:
```diff
@@ -1,63 +1,65 @@
 ; RUN: llc < %s -mtriple=aarch64-windows | FileCheck %s
 
 define dso_local i32 @func(ptr %g, i32 %a) "sign-return-address"="non-leaf" "sign-return-address-key"="b_key" {
+; CHECK-LABEL: func:
```
Nit: As this moves around the test patterns, it's hard to see how this changes the output. Can you precommit a test update that just autoregenerates the patterns with the old code, so this just is a small diff for the changed codegen?

MacDue commented on llvm/lib/Target/AArch64/AArch64PrologueEpilogue.cpp:
```diff
@@ -179,11 +179,23 @@ AArch64PrologueEpilogueCommon::convertCalleeSaveRestoreToSPPrePostIncDec(
   (void)Success;
   assert(Success && "unknown load/store opcode");
 
+  const auto *TRI = Subtarget.getRegisterInfo();
```
`RegInfo` is already a class member.
```suggestion
```

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64PrologueEpilogue.cpp:
```diff
@@ -179,11 +179,23 @@ AArch64PrologueEpilogueCommon::convertCalleeSaveRestoreToSPPrePostIncDec(
+  // a sequence like "sub sp, sp, #16; stp x19, lr, [sp]; sub sp, sp, #16",
+  // but we currently generate "sub sp, sp, #32; stp x19, lr, [sp, #16]". We
+  // could handle that here, but it's not clearly profitable; it saves up to
+  // 4 words of xdata, but it costs 2 instructions.)
```
One instruction in the prologue, one instruction in the epilogue.  I'll try to rephrase.

efriedma-quic committed:
Fix variable.

efriedma-quic committed:
Merge branch 'main' into aarch64-windows-more-pairing

efriedma-quic committed:
Fix comments about stack bump

We actually do already have code to split the stack bump... but only when optimizing for size.

efriedma-quic committed:
Adjust comment wording.

efriedma-quic reviewed (commented):
Addressed review comments.

I didn't realize that shouldCombineCSRLocalStackBump already has some code to deal with split stack bumps... but it only triggers with optsize.  So I guess we don't need to do anything there, unless we want to refine the existing heuristic.

efriedma-quic committed:
Update llvm/lib/Target/AArch64/AArch64PrologueEpilogue.cpp

Co-authored-by: Benjamin Maxwell <macdue@dueutil.tech>

efriedma-quic committed:
Merge branch 'main' into aarch64-windows-more-pairing

efriedma-quic committed:
Fix formatting.

mstorsjo commented:
If there's no more comments here, we should probably merge this?

efriedma-quic committed:
Merge branch 'main' into aarch64-windows-more-pairing
